sdl2promela/builtins.py

```diff
@@ -24,29 +31,127 @@ def __translate_writeln(
+    if (
+        not isinstance(parameters[0], promelamodel.VariableReference)
+        and not isinstance(parameters[0], promelamodel.MemberAccess)
+        and not isinstance(parameters[0], promelamodel.ArrayAccess)
+    ):
+        raise ValueError("Invalid type of parameter for length")
+
+    return (
+        MemberAccessBuilder()
+        .withUtypeReference(parameters[0])
+        .withMember(VariableReferenceBuilder("length").build())
+        .build()
+    )
+
+
+def __translate_present(
+    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Expression]
+):
+    if len(parameters) != 1:
+        raise ValueError(
+            f"Present built-in function expected 1 argument, but received {len(parameters)}."
+        )
+    if parameters[0] is None:
+        raise ValueError("Present queried entity is missing.")
+    if (
+        not isinstance(parameters[0], promelamodel.VariableReference)
+        and not isinstance(parameters[0], promelamodel.MemberAccess)
+        and not isinstance(parameters[0], promelamodel.ArrayAccess)
+    ):
+        raise ValueError("Invalid type of parameter for present")
+
+    return (
+        MemberAccessBuilder()
+        .withUtypeReference(parameters[0])
+        .withMember(VariableReferenceBuilder("selection").build())
+        .build()
+    )
+
+
+def __translate_to_enum(
+    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Expression]
+):
+    if len(parameters) != 2:
+        raise ValueError(
+            f"to_enum built-in function expected 2 arguments, but received {len(parameters)}."
+        )
+    return (
+        BinaryExpressionBuilder(promelamodel.BinaryOperator.SUBTRACT)
+        .withLeft(parameters[0])
+        .withRight(promelamodel.IntegerValue(1))
+        .build()
+    )
+
+
+def __translate_to_selector(
+    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Expression]
+):
+    print(f"To selector: call={call} parameters={parameters}")
     if len(parameters) != 2:
-        # Length requires 2 arguments: first one is the result, the second one is the queried entity
         raise ValueError(
-            f"Length built-in function expected 2 arguments (1 out and 1 in), but received {len(parameters)}."
+            f"to_selector built-in function expected 2 arguments, but received {len(parameters)}."
+        )
+    return (
+        BinaryExpressionBuilder(promelamodel.BinaryOperator.ADD)
+        .withLeft(parameters[0])
+        .withRight(promelamodel.IntegerValue(1))
+        .build()
+    )
+
+
+def __translate_num(
+    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Expression]
+):
+    if len(parameters) != 1:
+        raise ValueError(
+            f"num built-in function expected 2 arguments, but received {len(parameters)}."
+        )
+    return parameters[0]
+
+
+def __translate_val(
+    call: sdlmodel.ProcedureCall, parameters: List[promelamodel.Expression]
+):
+    if len(parameters) != 2:
```

Comment:
I wonder thether this parameter checking and exception throwing could be extracted into a single method _checkParameterCount(parameters, "function_name", expectedCount).
Also, in some cases they are checked for nulls... in some they are not. Why the inconsistency?

Reply:
changed.